### ISSUE TITLE
Cross-domain meta-learning for bug finding in the source codes with a small dataset

### ISSUE THREAD

**Issue by ZHEINEN001** [team3-cs373/resources#14]
### Title
Cross-domain meta-learning for bug finding in the source codes with a small dataset
### URL
https://doi.org/10.1145/3424954.3424957
### Summary 
Finding security holes and patching them is an effective way to increase security. The problem is that it takes immense manpower to audit or reverse engineer all the new software to check for security leaks. Machine learning would help the tech community scale better, but it is proving difficult to teach machines what bugs look like. The sample sizes are just too small or they have fake bugs in them which would confuse the machine. They use a hybrid neural network model which includes pre-trained language model and a meta-learned model. The meta model was better than the other, but overall the results were still worse than humans.
### Key Points 
1. code auditing and reverse engineering are good ways to find bugs
1. meta learned models use algorithms and a few examples to teach the machine.
1. the hybrid neural network is still worse than humans.
### Citation
@inproceedings{10.1145/3424954.3424957,
author = {Shin, Jongho},
title = {Cross-Domain Meta-Learning for Bug Finding in the Source Codes with a Small Dataset},
year = {2020},
isbn = {9781450375993},
publisher = {Association for Computing Machinery},
address = {New York, NY, USA},
url = {https://doi.org/10.1145/3424954.3424957},
doi = {10.1145/3424954.3424957},
abstract = {In terms of application security, detecting security vulnerabilities in prior and fixing them is one of the effective ways to prevent malicious activities. However, finding security bugs is highly reliant upon human experts due to its complexity. Therefore, source code auditing, one of the ways to find bugs, costs a lot, and the quality of auditing quite varies according to the performer. There have been many attempts to make automated systems for code auditing, but they have been suffered from huge false positives and false negatives.Meanwhile, machine learning technology is advancing dramatically in recent years, and it is outperforming humans in many tasks with high accuracy. Thus there have been lots of efforts to accommodate machine learning technology for security research. Most of the time, however, it is very difficult to obtain legitimate training data, and rarer often means more lethal in security. Therefore it is not easy to build reliable machine learning systems for security defects, and we are highly relying on human experts who can learn easily by a few examples.To overcome the obstacle, this paper proposes a deep neural network model for finding security bugs, which takes advantages of the recent developments in the machine learning technology; the language model adapted sub-word tokenization and self-attention based transformer from natural language processing for source code understanding, and a meta-learning technique from computer vision to overcome lack of legitimate vulnerability samples for the deep learning model. The model is also evaluated for finding DOM-based XSS bugs which is prevalent but hard to spot with traditional detection methods. The result shows that the model outperforms the baseline by 45% in the F1 score.},
booktitle = {Proceedings of the European Interdisciplinary Cybersecurity Conference},
articleno = {8},
numpages = {6},
keywords = {cyber security, deep neural network, vulnerability detection, few-shot learning, cross domain, meta-learning},
location = {Rennes, France},
series = {EICC 2020}
}
